python/pyspark/ml/stat/distribution.py

```diff
@@ -0,0 +1,267 @@
+
+    def logpdf(self,x):
+        """
+        Returns the log-density of this multivariate Gaussian at a point given by Vector x
+        """
+        assert (isinstance(x, Vector)), "x must be of Vector Type"
+        return float(self.__logpdf(x))
+
+    def __calculateCovarianceConstants(self):
+        """
+        Calculates distribution dependent components used for the density function
+        based on scipy multivariate library
+        refer https://github.com/scipy/scipy/blob/master/scipy/stats/_multivariate.py
+        tested with precision of 9 significant digits(refer testcase)
+        
+
+        """
+
+        try :
+            # pre-processing input parameters
+            # throws ValueError with invalid inputs
+            self.dim, self.mu, self.sigma = self.__process_parameters(None, self.mu, self.sigma)
+
+            # return the eigenvalues and eigenvectors 
+            # of a Hermitian or symmetric matrix.
+            # s =  eigen values
+            # u = eigen vectors
+            s, u = np.linalg.eigh(self.sigma)
+
+            #Singular values are considered to be non-zero only if 
+            #they exceed a tolerance based on machine precision, matrix size, and
+            #relation to the maximum singular value (same tolerance used by, e.g., Octave).
+
+            # calculation for machine precision
+            t = u.dtype.char.lower()
+            factor = {'f': 1E3, 'd': 1E6}
+            cond = factor[t] * np.finfo(t).eps
+
+            eps = cond * np.max(abs(s))
+
+            # checkng whether covariance matrix has any non-zero singular values
+            if np.min(s) < -eps:
+                raise ValueError
+
+            #computing the pseudoinverse
+            s_pinv = self.__pinv_1d(s, eps)
+
+            # prec_U ndarray
+            # A decomposition such that np.dot(prec_U, prec_U.T)
+            # is the precision matrix, i.e. inverse of the covariance matrix.
+            self.prec_U = np.multiply(u, np.sqrt(s_pinv))
+
+            #log_det_cov : float
+            #Logarithm of the determinant of the covariance matrix
+            self.log_det_cov = np.sum(np.log(s[s > eps]))
+
+        except ValueError :
+            raise ValueError("Covariance matrix has no non-zero singular values")
+
+    def __pdf(self,x):
+        """
+        Calculates density at point x using precomputed Constants
+        """
+        return np.exp(self.__logpdf(x))
+
+    def __logpdf(self,x) :
+        """
+        Calculates log-density at point x using precomputed Constants
+
+        x  Points at which to evaluate the log of the probability
+            density function
+        log_det_cov : float
+            Logarithm of the determinant of the covariance matrix
+
+        prec_U ndarray
+            A decomposition such that np.dot(prec_U, prec_U.T)
+            is the precision matrix, i.e. inverse of the covariance matrix.
+
+        """
+        x = self.__process_quantiles(x, self.dim)
```

Comment:
It is safe to assume x is 1-D. You have just asserted that it is a Vector above. (i.e do away with this)